Input completion

joshbax189 opened the pull request:
Adds completion, supporting both `company` and `completion-at-point` completion styles out-of-the-box.

Call either `completion-at-point` or `company-manual-begin` to get completions.

## Limitations
- does not complete keywords, e.g. "for", "function" etc (although this could be enabled within company)
- `completion-at-point` does not complete global values, i.e. when the prompt is empty

## How it works
At a high level, it sends the current user input followed by TAB to the REPL, then turns the response back into a list of candidates.

The completion responses are dispatched to listening callbacks, which contain their own state in closures. The callbacks represent a specific completion, e.g. "complete 'Array.' with a list of properties", or "clear the input". This async design means we avoid blocking while waiting for responses, so completion does not feel laggy.

Specifically, `js-comint--async-output-filter` is the dispatcher: it diverts output to a hidden buffer `js-comint--completion-buffer` when there are callbacks in `js-comint--completion-callbacks` and when the input and prompt positions match. Each of `js-comint--get-completion-async` and `js-comint--clear-input-async` are listener "factories", which send some input and register a callback to handle the expected response, implicitly hiding the output until they are completed.

redguardtoo commented:
You don't need require `company` explicitly. The standard practice is using `completing-read` api which supported by `company-mode`. Could you also squash your commits into one commit which is easier for my review?

joshbax189 committed:
Add completion feature

Squashed commit of the following:

commit 479eb4664d9f7325980bd61e344bd7e631aa57fd
Author: Josh Bax <joshbax189@gmail.com>
Date:   Tue Dec 3 16:41:38 2024 -0800

    Fix: use emacs 28 compat completion fns in test

commit 5870d05b9417e3a46f3806cc76089a051842fe6a
Author: Josh Bax <joshbax189@gmail.com>
Date:   Tue Dec 3 16:17:33 2024 -0800

    Fix: avoid Eask issue to unblock tests

commit 19446fd9f146e92500ed511ddd57d42219091f8f
Author: Josh Bax <joshbax189@gmail.com>
Date:   Tue Dec 3 10:51:17 2024 -0800

    Tidy comments and docstrings

commit af87be00c890c506b562eee473d97d71eca1cae8
Author: Josh Bax <joshbax189@gmail.com>
Date:   Tue Dec 3 10:41:40 2024 -0800

    Fix do not pass arguments to callback

commit ee0824b7ebeb3f7bef8830e368de3b670c2e34fa
Author: Josh Bax <joshbax189@gmail.com>
Date:   Tue Dec 3 15:23:16 2024 -0800

    Add Capf tests

    * test/capf-tests.el: C

commit 9fcb825af2fe384b822361d1d15ca6f142c7a627
Author: Josh Bax <joshbax189@gmail.com>
Date:   Mon Dec 2 17:50:10 2024 -0800

    Move tests into own folder and separate files

    This allows separate loading of company and CAPE

commit bd2616b48575fcb66fb5e438b273f23486ced759
Author: Josh Bax <joshbax189@gmail.com>
Date:   Tue Dec 3 15:17:17 2024 -0800

    Native capf

commit 5d67b6ba8594214ec963ef5f93cd94892eb8e6c7
Author: Josh Bax <joshbax189@gmail.com>
Date:   Wed Nov 27 11:22:20 2024 -0800

    Use string-match-p to support 28

commit 9f1241f22e0eee9f3cbdf52ea391ceee2696b33e
Author: Josh Bax <joshbax189@gmail.com>
Date:   Wed Nov 27 11:03:56 2024 -0800

    Cleanup hung callbacks when completing props of null object

commit 89375779aa38aae4990267187bd54243fd35755f
Author: Josh Bax <joshbax189@gmail.com>
Date:   Tue Nov 26 12:45:12 2024 -0800

    Quote string when used as regexp in callback

commit e276f26555b458790d865205b8ec88fd86de4433
Author: Josh Bax <joshbax189@gmail.com>
Date:   Tue Nov 26 11:46:38 2024 -0800

    Address fast-typing bug: extra input when callbacks overlap

commit 4353c89f0e263f664c09d531dab1e38ba30aa6f5
Author: Josh Bax <joshbax189@gmail.com>
Date:   Tue Nov 26 10:28:00 2024 -0800

    Use ert-async for tests with checks in callbacks

    Update test-company-complete-long-line

commit 44f2eacd1e030d372ef5dfa3c9e1e36e03bec98b
Author: Josh Bax <joshbax189@gmail.com>
Date:   Mon Nov 25 12:32:37 2024 -0800

    Fix matching bug in complete-substring

commit bb479f5f3663a33a59fea97e91bcfcc9e8302bf6
Author: Josh Bax <joshbax189@gmail.com>
Date:   Mon Nov 25 11:34:27 2024 -0800

    Remove unused completion string property

commit ea7a9fd68810059b4e9e0b51a5885dad8bf23d62
Author: Josh Bax <joshbax189@gmail.com>
Date:   Mon Nov 25 10:28:29 2024 -0800

    Fix: minor formatting change

commit 2ae491d162e49b6109ec74351f432fa50a873e22
Author: Josh Bax <joshbax189@gmail.com>
Date:   Mon Oct 28 22:35:16 2024 -0700

    Cleanup completion buffers

commit e07de7bb02114529b5abf962a50431b70002fa80
Author: Josh Bax <joshbax189@gmail.com>
Date:   Mon Oct 28 22:22:11 2024 -0700

    Pass NODE_REPL_MODE as argument due to custom env

    Env var has no effect for some reason

commit b649559fb0055a29b5e0b093b0420a21bd618608
Author: Josh Bax <joshbax189@gmail.com>
Date:   Mon Oct 28 22:19:17 2024 -0700

    Wiggle the cursor to produce a prompt and prevent hanging

    This makes most cases resolve with similar output too

commit 123c87dec4434a9a6e3aa87f2b26306367325c12
Author: Josh Bax <joshbax189@gmail.com>
Date:   Mon Oct 28 21:23:52 2024 -0700

    Add Company integration tests

commit 83f8a31b31fbb63780618ce48f20bc81eb79cf6f
Author: Josh Bax <joshbax189@gmail.com>
Date:   Mon Oct 28 19:37:40 2024 -0700

    Simplify process-completion-output

commit f80ca422a6991690489554a710128fa99c0016d1
Author: Josh Bax <joshbax189@gmail.com>
Date:   Mon Oct 28 19:24:50 2024 -0700

    Proactively clear partial completions

commit 672c28afe896cf8cc60f1c641dbc7b8cae9465e5
Author: Josh Bax <joshbax189@gmail.com>
Date:   Mon Oct 28 16:08:57 2024 -0700

    Complete long statements by taking latest substring

commit 3e6863ef2f0f9064bdbdae72ad349416a4c984ee
Author: Josh Bax <joshbax189@gmail.com>
Date:   Mon Oct 28 15:48:04 2024 -0700

    Don't complete at whitespace

commit 51ebba7051e852e1e3e7d36e0bdc92d81fa3fd42
Author: Josh Bax <joshbax189@gmail.com>
Date:   Fri Oct 25 14:48:24 2024 -0700

    Add unit tests for listener callback code

commit 41577969df45572ef4e6d7cc1b9e9a734f117f57
Author: Josh Bax <joshbax189@gmail.com>
Date:   Fri Oct 25 14:42:12 2024 -0700

    Recover from errors in company callbacks after completion

commit 0ed6d0be08ef334c710a2dcd0ab67d5d95e91098
Author: Josh Bax <joshbax189@gmail.com>
Date:   Fri Oct 25 14:12:50 2024 -0700

    Set TERM using env

    Remove with-environment-variables because it must be set within the
    comint call. Move other vars into the same location.

commit aa3821050361f608469924f6144b33e2f186244c
Author: Josh Bax <joshbax189@gmail.com>
Date:   Fri Oct 25 10:22:44 2024 -0700

    Failed callbacks should be removed

commit 5b3efeaf6bde14e3729e80bea24bfc48e67a367e
Author: Josh Bax <joshbax189@gmail.com>
Date:   Fri Oct 25 09:45:53 2024 -0700

    Factor out callback active test

commit 67208a6cf769d4250342d46e61b6298999f9dbd4
Author: Josh Bax <joshbax189@gmail.com>
Date:   Thu Oct 24 23:11:50 2024 -0700

    Add retry counter for method completion

commit 9433d55aea03aeea9a9167451425d7a1e4bbcc42
Author: Josh Bax <joshbax189@gmail.com>
Date:   Thu Oct 24 23:11:38 2024 -0700

    Factor out callback creation

commit dd0a5e5deac9608c692f356b7be9767df2650adc
Author: Josh Bax <joshbax189@gmail.com>
Date:   Thu Oct 24 13:08:07 2024 -0700

    Rewrite completions to use listener pattern callback

commit 83e3509a5d472d3b7b90c42daf394eb218a20813
Author: Josh Bax <joshbax189@gmail.com>
Date:   Thu Oct 24 13:02:45 2024 -0700

    Rename company-js-comint-backend => company-js-comint

    Matches other backend names

commit 74592d57755a51e88508f41707cc636bf09a475a
Author: Josh Bax <joshbax189@gmail.com>
Date:   Tue Oct 22 23:56:41 2024 -0700

    Don't use comint-send-string for invisible input

commit 4816ab3dce13bfc40fcf1dc0a880b883d9c247a0
Author: Josh Bax <joshbax189@gmail.com>
Date:   Tue Oct 22 23:50:09 2024 -0700

    Inline company definition in should-complete

commit c6cbbda5b43028aa6aa995b736dcda93429494ee
Author: Josh Bax <joshbax189@gmail.com>
Date:   Tue Oct 22 12:48:22 2024 -0700

    Use 28.1 compatible kill-matching-buffers in test

commit bb64f539258354b2365a965470a9ca1475e79749
Author: Josh Bax <joshbax189@gmail.com>
Date:   Tue Oct 22 12:36:52 2024 -0700

    Add should-complete unit tests

    Add Eask file

commit 0dc206a7be64f61ba8098f22a962be317d115df6
Author: Josh Bax <joshbax189@gmail.com>
Date:   Tue Oct 22 11:23:47 2024 -0700

    Add test macro with-new-js-comint-buffer and simplify tests

    Kills comint buffer between runs for safety

commit 054c09781d9c2c1711b4e56b822650cb44169944
Author: Josh Bax <joshbax189@gmail.com>
Date:   Sun Oct 20 11:42:37 2024 -0700

    Add unload function to remove company backend

commit 96cf5a7d7f45402d67523e52872c5ffed9099f9c
Author: Josh Bax <joshbax189@gmail.com>
Date:   Wed Oct 16 16:32:28 2024 -0700

    Unit tests for normal completion

commit 2721fd884b06eeb06e456bfb78a9c52c5c33d704
Author: Josh Bax <joshbax189@gmail.com>
Date:   Wed Oct 16 16:24:30 2024 -0700

    Use js-comint-get-process

commit f2f3d65c8729acbf585c3494ab0768d526161276
Author: Josh Bax <joshbax189@gmail.com>
Date:   Wed Oct 16 15:50:56 2024 -0700

    Do not take arguments for js-comint--discard-output when a callback

commit 7829facfcbf47b90714391df3bc5748ead32e438
Author: Josh Bax <joshbax189@gmail.com>
Date:   Wed Oct 16 15:41:30 2024 -0700

    Discard should apply to all output until control char

commit 6e274c8fe3c50247a6481b1d6b943e1fd1b8bbc6
Author: Josh Bax <joshbax189@gmail.com>
Date:   Wed Oct 16 15:35:53 2024 -0700

    Add unit tests for js-comint--process-completion-output

commit 5b50adad45d65b73b5480b8d0b589b965cac068c
Author: Josh Bax <joshbax189@gmail.com>
Date:   Wed Oct 16 13:32:57 2024 -0700

    Properly complete method names

commit 7f57d7618e43ed43e1d7a47b46cac22ed4825b59
Author: Josh Bax <joshbax189@gmail.com>
Date:   Wed Oct 16 11:12:29 2024 -0700

    Factor out completion processing logic

commit a98dc6e4947a085eacc82f68539ad72638e3f4a3
Author: Josh Bax <joshbax189@gmail.com>
Date:   Wed Oct 16 09:15:20 2024 -0700

    Simplify handler logic

commit f182a523725c6e62a5c1808c2141c191d7c31acf
Author: Josh Bax <joshbax189@gmail.com>
Date:   Tue Oct 15 15:30:25 2024 -0700

    Completion handler with debug messages

commit 18ff4f64a0426885b6e79cb6a130284a2082cdde
Author: Josh Bax <joshbax189@gmail.com>
Date:   Tue Oct 15 13:43:59 2024 -0700

    Filter completion output when a flag is set

commit 1cfedc9854d4ce6f99a3a14132764ef720737955
Author: Josh Bax <joshbax189@gmail.com>
Date:   Tue Oct 15 10:13:52 2024 -0700

    Get input for completion

commit 1352dbf197c4cc27ffdd82e57e01cc02171bf367
Author: Josh Bax <joshbax189@gmail.com>
Date:   Tue Oct 15 09:01:49 2024 -0700

    Add stub for company-js-comint-backend and init code

joshbax189 commented:
> Could you also squash your commits into one commit which is easier for my review?

Done

> The standard practice is using completing-read api which supported by company-mode

I think you mean `completion-at-point` not `completing-read`?

The reason I chose to add `company-js-comint` is because I designed the completion to use async functions. For example, see the signature `js-comint--get-completion-async (input-string callback)`: the completion results are used as an argument to `callback`. This already matches the company API.

`completion-at-point-function` is a function that takes no arguments and (synchronously) returns `(START END COLLECTION . PROPS)`. To convert `js-comint--get-completion-async` to a `completion-at-point` function I call the async function then poll for a result, see the implementation of `js-comint--capf` (note CAPF = completion-at-point function).

So, rather than having `company-capf` (async) call `js-comint--capf` (sync) call `js-comint--get-completion-async` (async), I just directly provide a company backend and enable it if company is loaded.

If company is not loaded, then `js-comint--capf` is used, so you can use completion with no extra packages installed.

redguardtoo commented:
Thanks.
